keanu-project/src/test/java/io/improbable/keanu/network/ModelCompositionTest.java

```diff
@@ -0,0 +1,206 @@
+package io.improbable.keanu.network;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import io.improbable.keanu.algorithms.variational.optimizer.gradient.GradientOptimizer;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.VertexId;
+import io.improbable.keanu.vertices.VertexLabel;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.DoubleProxyVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.ParetoVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.UniformVertex;
+
+public class ModelCompositionTest {
+
+    private BayesianNetwork innerNet;
+    private UniformVertex trueLocation;
+    private BayesianNetwork outerNet;
+    private DoubleVertex location;
+    private DoubleVertex size;
+    private DoubleVertex gaussian;
+    private DoubleVertex pareto;
+    private DoubleVertex gaussOutputVertex;
+
+    /*
+     * These tests run over a single graph within a graph.  The "inner" graph has a single Proxy input for the location
+     * a probabilistic value for size and two potential outputs:
+     *
+     * | Proxy "Location" |              | Uniform - Size |
+     *                     \            /
+     *     | Gaussian ("Output1") & Pareto ("Output2") |
+     *
+     * The Outer graph hooks up a Uniform to the "Location" input of the inner graph and takes the Output "Output1" as
+     * the single output it cares about.
+     */
+    @Before
+    public void setup() {
+
+        createInnerNet();
```

Comment:
This looks a bit odd - can you change `createInnerNet` to return a `BayesianNetwork` and do `innerNet = createInnerNet();`